rb/lib/selenium/webdriver/common/credential.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+# Licensed to the Software Freedom Conservancy (SFC) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The SFC licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#
+# A credential stored in a virtual authenticator.
+# @see https://w3c.github.io/webauthn/#credential-parameters
+#
+
+module Selenium
+  module WebDriver
+    class Credential
+      def initialize(*args)
+        @id = args[0]
+        @is_resident_credential = args[1]
+        @rp_id = args[2]
+        @user_handle = args[3]
+        @private_key = args[4]
+        @sign_count = args[5]
+      end
+
+      attr_reader :id, :is_resident_credential, :rp_id, :private_key, :sign_count
+
+      def user_handle
+        return @user_handle unless @user_handle.nil?
+
+        nil
+      end
+
+      def create_resident_credential(id, rp_id, user_handle, private_key, sign_count)
+        Credential.new(id, true, rp_id, user_handle, private_key, sign_count)
+      end
+
+      def create_non_resident_credential(id, rp_id, private_key, sign_count)
+        Credential.new(id, false, rp_id, nil, private_key, sign_count)
+      end
+
+      def to_dict
+        credential_data = {
+          credentialId: Base64.urlsafe_encode64(@id&.pack('C*')),
+          isResidentCredential: @is_resident_credential,
+          rpId: @rp_id,
+          privateKey: Base64.urlsafe_encode64(@private_key),
+          signCount: @sign_count
+        }
+
+        credential_data[:userHandle] = Base64.urlsafe_encode64(@user_handle&.pack('C*')) unless user_handle.nil?
+        credential_data
+      end
+
+      def from_dict(data)
```

Comment:
Let's make this a class method so that it would return an instance of credential class:

```suggestion
      def self.from_dict(data)
```

Reply:
Agree on the class method, but I'd suggest `#from_json` instead of `#from_dict` since that's what we're actually converting from, and we're using that method name at least one other place

```suggestion
      def self.from_json(data)
```

Reply:
Seems right! Thank you @p0deje and @titusfortner. Made the change.